pkg/util/jsonpath: add variable, array indexing support for jsonpath parser

normanchenn opened the pull request:
This PR adds a bunch of small changes to the jsonpath parser.
- The AST is refactored so that the nodes satisfy one interface
- Key accessors are now wrapped in double quotes to emulate postgres
- Add support for variables to be parsed within jsonpath queries
- Refactored `scanner.scanNumber` so sql and plpgsql scanners can use the same logic
- Add support for array indexes to be parsed within jsonpath queries
- Add support for array ranges to be parsed within jsonpath queries
- Add support for array unions to be parsed within jsonpath queries

Epic: None
Release note (sql change): Add support for variables, array indexing within jsonpath parser.

normanchenn committed:
pkg/util/jsonpath: refactor jsonpath AST

Previously, the jsonpath AST consisted of nodes that satisfied either
`Expr` or `Accessor` interfaces. This would become harder to maintain
once we add more functionalities to the language. In this commit, the
jsonpath AST has been refactored so that there is a `Jsonpath` struct
representing a jsonpath query, and all the other nodes within the AST
satisfy a `Path` interface.

Epic: None
Release note: None

normanchenn committed:
pkg/sql/scanner: support double-quoted key accessors in jsonpath

Previously, the jsonpath parser did not support double-quoted key
accessors (ex. `$."key1"`). This commit adds this functionality within
the jsonpath parser. This is helpful for key accessors that start
with a digit (`$.1key`), which currently isn't supported without the
double quotes.

Epic: None
Release note (sql change): Support double-quoted key accessors within
jsonpath (`SELECT '$."1key"."key2"'::JSONPATH;`).

normanchenn committed:
pkg/util/jsonpath: add quotes around jsonpath keys

Previously, the formatted jsonpath text wouldn't wrap the key names in
quotes. To match better with postgres, this commit adds the double
quotes around the key names.

Epic: None
Release note: None

normanchenn committed:
pkg/util/jsonpath: support variables within jsonpath parser

This commit adds parsing for variables within jsonpath. Variables are
denoted as `$name`, and are injected into the jsonpath query as an
optional parameter within `jsonb_path_query`.

Epic: None
Release note: None

normanchenn committed:
pkg/util/jsonpath: support jsonpath array indexing

This commit adds simple array indexing for jsonpath queries. Integer
indexes are supported, while negative and float indexes which are
supported within postgres are not yet supported.

Epic: None
Release note (sql change): Support jsonpath array integer indexing
(ex. `$.a[1]`).

normanchenn committed:
pkg/util/jsonpath: support jsonpath array ranges

This commits add functionality for specifying array ranges within
jsonpath queries (`$.a[1 to 3]`).

Epic: None
Release note (sql change): Support jsonpath array ranges (ex. `$.a[1 to
3]`).

normanchenn committed:
pkg/util/jsonpath: support jsonpath array unions

This commit adds functionality for jsonpath array unions (`$.a[1, 2 to 4,
7, 8]`).

Epic: None
Release note (sql change): Support jsonpath array unions (ex `$.a[1, 2 to 4,
7, 8]`).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142241)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice! I just have one suggestion to avoid some code duplication.

> Reviewed 2 of 2 files at r1, 3 of 3 files at r2, 5 of 5 files at r3, 4 of 4 files at r4, 6 of 6 files at r5, 3 of 3 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-:-OKXuSxx8TvAoqyzcIps:b-d85hov)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn)

___
*[`pkg/util/jsonpath/parser/jsonpath.y` line 305 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXu5gA1JY3MdI5NzRc:-OKXu5gA1JY3MdI5NzRd:b2cqdld) ([raw file](https://github.com/cockroachdb/cockroach/blob/7157f3656d2d39daa2eb0f116bc5832a3d0ae119/pkg/util/jsonpath/parser/jsonpath.y#L305)):*
> ```Yacc
> 
> unreserved_keyword:
>   STRICT
> ```

nit: it'd be nice to keep this list lexicographically ordered.
___
*[`pkg/util/jsonpath/parser/testdata/jsonpath` line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXSsOaCkfzf2pLr4Ij:-OKXSsOb1jhxPw50_pDy:b7n7d4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bbbce58d41a2cc47429c30e63c3c2c1a7713ab/pkg/util/jsonpath/parser/testdata/jsonpath#L163)):*
> ```
> #  $.1a
> #  ----
> #  $.1a
> ```

nit: seems like missing newline at the end of the file in the second commit.
___
*[`pkg/sql/scanner/jsonpath_scan.go` line 82 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXrN4iDj0wp1rRTHo0:-OKXrN4iDj0wp1rRTHo1:bl7lntv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d385a4afc9e7d1d8bc9a781124df77a2d944c0/pkg/sql/scanner/jsonpath_scan.go#L82)):*
> ```Go
> 
> // scanNumber is similar to Scanner.scanNumber, but uses Jsonpath tokens.
> func (s *JSONPathScanner) scanNumber(lval ScanSymType, ch int) {
> ```

Hm, we now will have almost an exact copy of this function in three `scanNumber` implementations. I think it might be worth extracting a shared helper method `scanNumberImpl` since the only difference I see is the token ID that is used for `ERROR`, `FCONST`, and `ICONST`, so we could pass those as arguments to the helper.
___
*[`pkg/sql/scanner/jsonpath_scan.go` line 131 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXohRd0CHuT9VC9A0n:-OKXohRd0CHuT9VC9A0o:btzmesp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d385a4afc9e7d1d8bc9a781124df77a2d944c0/pkg/sql/scanner/jsonpath_scan.go#L131)):*
> ```Go
> 				s.pos++
> 			}
> 			ch = s.peek()
> ```

nit: I find this way a bit easier to follow:
```
			ch = s.peek()
			if ch == '-' || ch == '+' {
				s.pos++
				ch = s.peek()
			}
```
Consider adding another commit in which all three scanners are cleaned up this way, or including this change into the commit with a shared helper I'm suggesting above.
___
*[`pkg/sql/scanner/jsonpath_scan.go` line 142 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXr_zt8-QJi3lbkXs4:-OKXr_zt8-QJi3lbkXs5:bxl7gmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d385a4afc9e7d1d8bc9a781124df77a2d944c0/pkg/sql/scanner/jsonpath_scan.go#L142)):*
> ```Go
> 	}
> 
> 	// Disallow identifier after numerical constants e.g. "124foo".
> ```

@DrewKimball do you remember why 4b1ac2ca82a9217e49711eb5021826db5edc9e01 omitted this check from the PLpgSQL variant?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142241)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn and @yuzefovich)

___
*[`pkg/sql/scanner/jsonpath_scan.go` line 142 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXr_zt8-QJi3lbkXs4:-OKYBt21CoFepg2xEvta:bkzb4y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d385a4afc9e7d1d8bc9a781124df77a2d944c0/pkg/sql/scanner/jsonpath_scan.go#L142)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

@DrewKimball do you remember why 4b1ac2ca82a9217e49711eb5021826db5edc9e01 omitted this check from the PLpgSQL variant?
</blockquote></details>

It looks like this check was added in #112021, after #104117. FWIW, I believe any numeric constants in a PL/pgSQL routine will end up passing through the SQL parser, so we should still  hit the check. Example:
```
root@localhost:26257/defaultdb> create function f() returns int language plpgsql as $$ begin for i in 1 .. 2he loop raise notice 'here'; end loop; end $$;
ERROR: at or near "loop": syntax error: lexical error: trailing junk after numeric literal at or near "2h"
SQLSTATE: 42601
DETAIL: source SQL:
SET ROW (2he )
         ^
--
source SQL:
begin for i in 1 .. 2he loop raise notice 'here'; end loop; end
                        ^
HINT: try \h SET SESSION
```


<!-- Sent from Reviewable.io -->

normanchenn committed:
sql/scanner: extract scanNumber implementation from token IDs

Previously, `scanNumber` scanning logic was duplicated across SQL and
PLpgSQL scanners, because they have different token ID values. This
commit extracts the common logic into a shared helper method which now
both scanners use.

Epic: None
Release note: None

normanchenn reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142241)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn and @yuzefovich)

___
*[`pkg/sql/scanner/jsonpath_scan.go` line 82 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXrN4iDj0wp1rRTHo0:-OKYMG1kCikwHXdoyiTU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d385a4afc9e7d1d8bc9a781124df77a2d944c0/pkg/sql/scanner/jsonpath_scan.go#L82)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, we now will have almost an exact copy of this function in three `scanNumber` implementations. I think it might be worth extracting a shared helper method `scanNumberImpl` since the only difference I see is the token ID that is used for `ERROR`, `FCONST`, and `ICONST`, so we could pass those as arguments to the helper.
</blockquote></details>

Done.
___
*[`pkg/sql/scanner/jsonpath_scan.go` line 131 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXohRd0CHuT9VC9A0n:-OKYMIb94mR8hyLrQQn-:byqbij7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d385a4afc9e7d1d8bc9a781124df77a2d944c0/pkg/sql/scanner/jsonpath_scan.go#L131)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I find this way a bit easier to follow:
```
			ch = s.peek()
			if ch == '-' || ch == '+' {
				s.pos++
				ch = s.peek()
			}
```
Consider adding another commit in which all three scanners are cleaned up this way, or including this change into the commit with a shared helper I'm suggesting above.
</blockquote></details>

Added this change in the commit refactoring the `scanNumber`.
___
*[`pkg/sql/scanner/jsonpath_scan.go` line 142 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-OKXr_zt8-QJi3lbkXs4:-OKYMYYn5uRfHFHMeX3P:bymlxod) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d385a4afc9e7d1d8bc9a781124df77a2d944c0/pkg/sql/scanner/jsonpath_scan.go#L142)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It looks like this check was added in #112021, after #104117. FWIW, I believe any numeric constants in a PL/pgSQL routine will end up passing through the SQL parser, so we should still  hit the check. Example:
```
root@localhost:26257/defaultdb> create function f() returns int language plpgsql as $$ begin for i in 1 .. 2he loop raise notice 'here'; end loop; end $$;
ERROR: at or near "loop": syntax error: lexical error: trailing junk after numeric literal at or near "2h"
SQLSTATE: 42601
DETAIL: source SQL:
SET ROW (2he )
         ^
--
source SQL:
begin for i in 1 .. 2he loop raise notice 'here'; end loop; end
                        ^
HINT: try \h SET SESSION
```
</blockquote></details>

I consolidated `scanNumber` to one implementation, so now PLpgSQL will also have that check.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Great, thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r8, 3 of 3 files at r9, 5 of 5 files at r10, 4 of 4 files at r11, 2 of 2 files at r12, 6 of 6 files at r13, 3 of 3 files at r14, 3 of 3 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142241#-:-OKZSEe_7WSYrIfYa_JV:b-napfio)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @normanchenn)



<!-- Sent from Reviewable.io -->

normanchenn commented:
TFTR!

bors r+
